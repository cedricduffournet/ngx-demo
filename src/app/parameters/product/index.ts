export * from './parameter-product.component';
