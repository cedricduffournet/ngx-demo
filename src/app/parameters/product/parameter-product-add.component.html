<app-product-add-view></app-product-add-view>
