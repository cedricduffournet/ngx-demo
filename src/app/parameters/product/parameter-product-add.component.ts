import { ChangeDetectionStrategy, Component, OnInit } from '@angular/core';

@Component({
  selector: 'app-parameter-product-add',
  templateUrl: './parameter-product-add.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class ParameterProductAddComponent implements OnInit {
  constructor() {}

  ngOnInit() {}
}
