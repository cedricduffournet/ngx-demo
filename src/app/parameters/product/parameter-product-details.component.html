<app-product-details-view></app-product-details-view>
