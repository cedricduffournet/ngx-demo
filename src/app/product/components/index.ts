export * from './product-add/product-add.component';
export * from './product-update/product-update.component';
export * from './product-delete/product-delete.component';
export * from './product-form/product-form.component';
export * from './product-item/product-item.component';
export * from './product-items/product-items.component';
export * from './product-details/product-details.component';
