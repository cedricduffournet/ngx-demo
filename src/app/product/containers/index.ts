export * from './product-add-view/product-add-view.component';
export * from './product-update-modal/product-update-modal.component';
export * from './product-delete-modal/product-delete-modal.component';
export * from './product-list-view/product-list-view.component';
export * from './product-selected/product-selected.component';
export * from './product-details-view/product-details-view.component';
