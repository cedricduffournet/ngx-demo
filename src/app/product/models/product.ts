import { schema } from 'normalizr';

export interface Product {
  id: number;
  name: string;
}

export const productSchema = new schema.Entity('products');
