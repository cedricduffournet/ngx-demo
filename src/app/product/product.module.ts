import { NgModule, ComponentFactoryResolver } from '@angular/core';

import { RouterModule } from '@angular/router';
import { EffectsModule } from '@ngrx/effects';
import { StoreModule } from '@ngrx/store';

import { SharedModule } from '@app/shared/shared.module';
import { ProductEffects } from '@app/product/state/effects';
import { ProductFacade } from '@app/product/state/product.facade';
import {
  ProductAddComponent,
  ProductUpdateComponent,
  ProductDeleteComponent,
  ProductFormComponent,
  ProductItemComponent,
  ProductItemsComponent
} from '@app/product/components';
import {
  ProductAddModalComponent,
  ProductUpdateModalComponent,
  ProductDeleteModalComponent,
  ProductListViewComponent
} from '@app/product/containers';
import { reducers } from '@app/product/state/reducers';
import { CoalescingComponentFactoryResolver } from '@app/coalescing-component-factory-resolver.service';

@NgModule({
  imports: [
    RouterModule,
    SharedModule,
    EffectsModule.forFeature([ProductEffects]),
    StoreModule.forFeature('products', reducers)
  ],
  declarations: [
    ProductListViewComponent,
    ProductAddComponent,
    ProductUpdateComponent,
    ProductDeleteComponent,
    ProductFormComponent,
    ProductItemComponent,
    ProductItemsComponent,
    ProductAddModalComponent,
    ProductUpdateModalComponent,
    ProductDeleteModalComponent
  ],
  entryComponents: [
    ProductUpdateModalComponent,
    ProductAddModalComponent,
    ProductDeleteModalComponent
  ],
  exports: [ProductListViewComponent],
  providers: [ProductFacade]
})
export class ProductModule {
  // see https://github.com/angular/angular/issues/14324
  constructor(
    coalescingResolver: CoalescingComponentFactoryResolver,
    localResolver: ComponentFactoryResolver
  ) {
    coalescingResolver.registerResolver(localResolver);
  }
}
