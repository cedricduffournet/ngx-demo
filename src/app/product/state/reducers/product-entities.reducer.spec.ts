import { reducer } from '@app/product/state/reducers/product-entities.reducer';
import * as fromProducts from '@app/product/state/reducers/product-entities.reducer';

import {
  ProductListViewActions,
  ProductUpdateModalActions,
  ProductApiActions
} from '@app/product/state/actions';
import { Product } from '@app/product/models/product';

describe('ProductEntitiesReducer', () => {
  describe('undefined action', () => {
    it('should return the default state', () => {
      const result = reducer(undefined, {} as any);

      expect(result).toMatchSnapshot();
    });
  });

  describe('LOAD_SUCCESS', () => {
    it('should return state with product entities', () => {
      const products = {
        entities: {
          products: {
            1: {
              name: 'Name1'
            } as Product,
            2: {
              name: 'Name2'
            } as Product
          }
        },
        result: [1, 2]
      };

      const meta = {
      };

      const action = ProductApiActions.loadProductSuccess({
        products,
        meta
      });
      const result = reducer(fromProducts.INITIAL_STATE, action);

      expect(result).toMatchSnapshot();
    });
  });

  describe('UPDATE', () => {
    const id = 1;
    const product = {
      name: 'NameUpdated'
    } as Product;
    const data = { id, product };

    it('should set updating to true', () => {
      const action = ProductUpdateModalActions.updateProduct({ data });
      const result = reducer(fromProducts.INITIAL_STATE, action);

      expect(result).toMatchSnapshot();
    });
  });

  describe('UPDATE_SUCCESS', () => {
    it('should update product id 1, upated should be true and updating false', () => {
      const products = {
        1: {
          id: 1,
          name: 'Name 1'
        } as Product,
        2: {
          id: 2,
          name: 'Name 2'
        } as Product
      };
      const initialState = {
        ...fromProducts.INITIAL_STATE,
        entities: products
      };

      const product = {
        entities: {
          products: {
            1: {
              name: 'Name 1 updated'
            }
          }
        },
        result: 1
      };
      const action = ProductApiActions.updateProductSuccess({ product });
      const result = reducer(initialState, action);

      expect(result).toMatchSnapshot();
    });
  });

  describe('UPDATE_FAILURE & SHOW_MODAL_UPDATE', () => {
    const initialState = {
      ...fromProducts.INITIAL_STATE,
      updating: true,
      updated: true,
      entities: {
        1: {
          id: 1,
          name: 'Name'
        } as Product
      }
    };

    it('should set updating to false', () => {
      const error = {
        message: 'error'
      };
      const action = ProductApiActions.updateProductFailure({ error });
      const result = reducer(initialState, action);

      expect(result).toMatchSnapshot();
    });

    it('should set updating to false, updated to false when show modal', () => {
      const action = ProductListViewActions.showUpdateProductModal();
      const result = reducer(initialState, action);

      expect(result).toMatchSnapshot();
    });
  });

  describe('ADD_SUCCESS', () => {
    const products = {
      1: {
        id: 1,
        name: 'Name 1'
      } as Product,
      2: {
        id: 2,
        name: 'Name 2'
      } as Product
    };
    const initialState = {
      ...fromProducts.INITIAL_STATE,
      entities: products
    };

    const product = {
      entities: {
        products: {
          3: {
            name: 'Name 3'
          }
        }
      },
      result: 3
    };

    it('should add a new product id 3', () => {
      const action = ProductApiActions.addProductSuccess({ product });
      const result = reducer(initialState, action);

      expect(result).toMatchSnapshot();
    });
  });

  describe('SELECT', () => {
    const products = {
      1: {
        id: 1,
        name: 'Name 1'
      } as Product,
      2: {
        id: 2,
        name: 'Name 2'
      } as Product
    };
    const initialState = {
      ...fromProducts.INITIAL_STATE,
      entities: products
    };

    const id = 1;

    it('should set selected product (id 1)', () => {
      const action = ProductListViewActions.selectProduct({ id });
      const result = reducer(initialState, action);

      expect(result).toMatchSnapshot();
    });
  });

  describe('Selectors', () => {
    describe('getEntities', () => {
      it('should return entities', () => {
        const result = fromProducts.getEntities({
          ...fromProducts.INITIAL_STATE,
          entities: {
            1: {
              id: 1,
              name: 'Name 1'
            } as Product,
            2: {
              id: 2,
              name: 'Name 2'
            } as Product
          }
        });

        expect(result).toMatchSnapshot();
      });
    });

    describe('getSelectedId', () => {
      it('should return selected id (1)', () => {
        const result = fromProducts.getSelectedId({
          ...fromProducts.INITIAL_STATE,
          selectedId: 1
        });

        expect(result).toMatchSnapshot();
      });
    });

    describe('getUpdating', () => {
      it('should return updating true', () => {
        const result = fromProducts.getUpdating({
          ...fromProducts.INITIAL_STATE,
          updating: true
        });

        expect(result).toMatchSnapshot();
      });
    });

    describe('getUpdated', () => {
      it('should return updated true', () => {
        const result = fromProducts.getUpdated({
          ...fromProducts.INITIAL_STATE,
          updated: true
        });

        expect(result).toMatchSnapshot();
      });
    });
  });
});
