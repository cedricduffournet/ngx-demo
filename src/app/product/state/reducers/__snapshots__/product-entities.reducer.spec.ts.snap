// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductEntitiesReducer ADD_SUCCESS should add a new product id 3 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "id": 1,
      "name": "Name 1",
    },
    "2": Object {
      "id": 2,
      "name": "Name 2",
    },
    "3": Object {
      "name": "Name 3",
    },
  },
  "selectedId": null,
  "updated": false,
  "updating": false,
}
`;

exports[`ProductEntitiesReducer LOAD_SUCCESS should return state with product entities 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "name": "Name1",
    },
    "2": Object {
      "name": "Name2",
    },
  },
  "selectedId": null,
  "updated": false,
  "updating": false,
}
`;

exports[`ProductEntitiesReducer SELECT should set selected product (id 1) 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "id": 1,
      "name": "Name 1",
    },
    "2": Object {
      "id": 2,
      "name": "Name 2",
    },
  },
  "selectedId": 1,
  "updated": false,
  "updating": false,
}
`;

exports[`ProductEntitiesReducer Selectors getEntities should return entities 1`] = `
Object {
  "1": Object {
    "id": 1,
    "name": "Name 1",
  },
  "2": Object {
    "id": 2,
    "name": "Name 2",
  },
}
`;

exports[`ProductEntitiesReducer Selectors getSelectedId should return selected id (1) 1`] = `1`;

exports[`ProductEntitiesReducer Selectors getUpdated should return updated true 1`] = `true`;

exports[`ProductEntitiesReducer Selectors getUpdating should return updating true 1`] = `true`;

exports[`ProductEntitiesReducer UPDATE should set updating to true 1`] = `
Object {
  "entities": Object {},
  "selectedId": null,
  "updated": false,
  "updating": true,
}
`;

exports[`ProductEntitiesReducer UPDATE_FAILURE & SHOW_MODAL_UPDATE should set updating to false 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "id": 1,
      "name": "Name",
    },
  },
  "selectedId": null,
  "updated": false,
  "updating": false,
}
`;

exports[`ProductEntitiesReducer UPDATE_FAILURE & SHOW_MODAL_UPDATE should set updating to false, updated to false when show modal 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "id": 1,
      "name": "Name",
    },
  },
  "selectedId": null,
  "updated": false,
  "updating": false,
}
`;

exports[`ProductEntitiesReducer UPDATE_SUCCESS should update product id 1, upated should be true and updating false 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "name": "Name 1 updated",
    },
    "2": Object {
      "id": 2,
      "name": "Name 2",
    },
  },
  "selectedId": null,
  "updated": true,
  "updating": false,
}
`;

exports[`ProductEntitiesReducer undefined action should return the default state 1`] = `
Object {
  "entities": Object {},
  "selectedId": null,
  "updated": false,
  "updating": false,
}
`;
